pkg/service/service.go

```diff
@@ -797,8 +797,25 @@ func (d *DynamicCRD) GetServiceNameFromCRD() (string, error) {
 	return "", fmt.Errorf("couldn't find metadata.name in the yaml; provide a name for the service")
 }
 
+// AddComponentLabelsToCRD appends odo labels to CRD if "labels" field already exists in metadata; else creates labels
+func (d *DynamicCRD) AddComponentLabelsToCRD(labels map[string]string) {
+	metaMap := d.OriginalCRD["metadata"].(map[string]interface{})
+
+	for k := range metaMap {
+		if k == "labels" {
+			metaLabels := metaMap["labels"].(map[string]interface{})
```

Comment:
I am not very clear on the `.(Type)` thing, but I assume you are looking for a value matching `.(map[string]interface{})` type. In that case, wouldn't you want to accept the `labels` arg of the same type instead of `.(map[string]string)`? I understand string itself is an interface but so are other types.

Reply:
I won't claim to be an expert here. But it *might* (because I haven't tested/checked) fail if there was already a label set in the manifest like below:
```yaml
- kubernetes:
    inlined: |
      apiVersion: etcd.database.coreos.com/v1beta2
      kind: EtcdCluster
      metadata:
        annotations:
          etcd.database.coreos.com/scope: clusterwide
        name: example
        
        labels:
          dharmit: shah
          intValue: 1    <------  integer value
      spec:
        size: 3
        version: 3.2.13
  name: example
```

Reply:
I tried with `map[string]string` and it panicked when I did `odo push`:
```sh
Creating Kubernetes resources for component node-todo
panic: interface conversion: interface {} is map[string]interface {}, not map[string]string

goroutine 1 [running]:
github.com/openshift/odo/pkg/service.(*DynamicCRD).AddComponentLabelsToCRD(0xc000825348, 0xc000ad1140)
	/home/dshah/go/src/github.com/openshift/odo/pkg/service/service.go:806 +0x326
github.com/openshift/odo/pkg/service.CreateServiceFromKubernetesInlineComponents(0xc0004d63f0, 0xc000b75320, 0x1, 0x1, 0xc000ad1140, 0x0, 0x1f26dba, 0xa, 0x0, 0x0)
	/home/dshah/go/src/github.com/openshift/odo/pkg/service/service.go:854 +0x147
github.com/openshift/odo/pkg/devfile/adapters/kubernetes/component.Adapter.Push(0xc0004d63f0, 0x22421e0, 0xc000478a40, 0x21f2400, 0xc000478aa0, 0x22164a0, 0xc000478b20, 0x22164e0, 0xc000478b80, 0x21f2480, ...)
	/home/dshah/go/src/github.com/openshift/odo/pkg/devfile/adapters/kubernetes/component/adapter.go:187 +0x805
github.com/openshift/odo/pkg/devfile/adapters/kubernetes.Adapter.Push(0x224a140, 0xc00074b2b0, 0xc0008a7a10, 0x26, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/dshah/go/src/github.com/openshift/odo/pkg/devfile/adapters/kubernetes/adapter.go:37 +0x79
github.com/openshift/odo/pkg/odo/cli/component.(*PushOptions).devfilePushInner(0xc00062b8c0, 0xc0001180e0, 0x0)
	/home/dshah/go/src/github.com/openshift/odo/pkg/odo/cli/component/devfile.go:139 +0x5b5
github.com/openshift/odo/pkg/odo/cli/component.(*PushOptions).DevfilePush(0xc00062b8c0, 0xc, 0x1)
	/home/dshah/go/src/github.com/openshift/odo/pkg/odo/cli/component/devfile.go:56 +0x2f
github.com/openshift/odo/pkg/odo/cli/component.(*PushOptions).Run(0xc00062b8c0, 0x0, 0x0)
	/home/dshah/go/src/github.com/openshift/odo/pkg/odo/cli/component/push.go:256 +0x51
github.com/openshift/odo/pkg/odo/genericclioptions.GenericRun(0x2216960, 0xc00062b8c0, 0xc00032b340, 0x2fdc758, 0x0, 0x0)
	/home/dshah/go/src/github.com/openshift/odo/pkg/odo/genericclioptions/runnable.go:93 +0x2b0
github.com/openshift/odo/pkg/odo/cli/component.NewCmdPush.func1(0xc00032b340, 0x2fdc758, 0x0, 0x0)
	/home/dshah/go/src/github.com/openshift/odo/pkg/odo/cli/component/push.go:281 +0x5e
github.com/spf13/cobra.(*Command).execute(0xc00032b340, 0x2fdc758, 0x0, 0x0, 0xc00032b340, 0x2fdc758)
	/home/dshah/go/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:854 +0x2c2
github.com/spf13/cobra.(*Command).ExecuteC(0xc000291600, 0x205dbc0, 0xc00010f980, 0x1)
	/home/dshah/go/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:958 +0x375
github.com/spf13/cobra.(*Command).Execute(...)
	/home/dshah/go/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:895
main.main()
	/home/dshah/go/src/github.com/openshift/odo/cmd/odo/odo.go:68 +0x368

```
Reason being that `metaMap` variable, that we're using to fetch the labels from, is a `map[string]interface{}`.

Reply:
> But it might (because I haven't tested/checked) fail if there was already a label set in the manifest like below:

But if that is the case, like the one in your manifest, wouldn't we want to add similar labels of types other than a string?

>I tried with map[string]string and it panicked when I did odo push:

I meant wouldn't you want to match the `labels` arg's type to  `map[string]interface{}` type?

Reply:
> wouldn't we want to add similar labels of types other than a string?

But we want add only those labels which are generated by `GetLabels` function. And it's quite specific about what we want to add.

>  wouldn't you want to match the labels arg's type to map[string]interface{} type?

The `labels` arg here is simply what we get from `componentlabels.GetLabels` and it returns a `map[string]string`.